ext/soap/soap.c

```diff
@@ -565,16 +574,15 @@ PHP_METHOD(SoapHeader, __construct)
 		add_property_zval(this_ptr, "data", data);
 	}
 	add_property_bool(this_ptr, "mustUnderstand", must_understand);
-	if (actor == NULL) {
-	} else if (Z_TYPE_P(actor) == IS_LONG &&
-	  (Z_LVAL_P(actor) == SOAP_ACTOR_NEXT ||
-	   Z_LVAL_P(actor) == SOAP_ACTOR_NONE ||
-	   Z_LVAL_P(actor) == SOAP_ACTOR_UNLIMATERECEIVER)) {
-		add_property_long(this_ptr, "actor", Z_LVAL_P(actor));
-	} else if (Z_TYPE_P(actor) == IS_STRING && Z_STRLEN_P(actor) > 0) {
-		add_property_stringl(this_ptr, "actor", Z_STRVAL_P(actor), Z_STRLEN_P(actor));
-	} else {
-		php_error_docref(NULL, E_WARNING, "Invalid actor");
+
+	if (!actor_is_null) {
+		if (actor_str && ZSTR_LEN(actor_str) > 2) {
```

Comment:
These checks should be split, so we don't check actor_long if actor_str is set but too short. I think you might want to split the "Invalid actor" error condition into two separate ValueErrors, one that says that it's too short and the other that it has to be one of SOAP_ACTOR.

Reply:
I tried my best to rearrange the conditions sensibly.